File: mnist_project/nn_mnist.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	numSamples = 1000
	numPixels  = 400
	hidden     = 64
	epochs     = 10
	lr         = 0.01
)

// ─────────────────────────────────────────
//  LOAD FLAT CSV
//  Same strategy as all other languages:
//  stream every value, skip header if present
// ─────────────────────────────────────────
func loadFlatCSV(path string, count int, hasHeader bool) []float64 {
	data := make([]float64, 0, count)

	file, err := os.Open(path)
	if err != nil {
		fmt.Printf("Error: Could not open %s\n", path)
		return data
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 1024*1024*64), 1024*1024*64)

	if hasHeader {
		scanner.Scan() 
	}

	for scanner.Scan() && len(data) < count {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		for _, part := range parts {
			part = strings.TrimSpace(part)
			if part == "" {
				continue
			}
			val, err := strconv.ParseFloat(part, 64)
			if err == nil {
				data = append(data, val)
			}
			if len(data) >= count {
				return data
			}
		}
	}
	return data
}

// ─────────────────────────────────────────
//  ACTIVATIONS
// ─────────────────────────────────────────
func relu(x float64) float64      { if x > 0.0 { return x }; return 0.0 }
func reluD(x float64) float64     { if x > 0.0 { return 1.0 }; return 0.0 }
func sigmoid(x float64) float64   { return 1.0 / (1.0 + math.Exp(-x)) }

func main() {
	totalStart := time.Now()

	// ─────────────────────────────────────
	//  LOAD DATA AND WEIGHTS
	// ─────────────────────────────────────
	fmt.Println("Loading dataset and weights...")
	loadStart := time.Now()

	
	data := loadFlatCSV(
		"mnist_project/data/Mnist_Binary_0_vs_1_1000_20x20.csv",
		numSamples*(numPixels+1), true)

	W1 := loadFlatCSV("mnist_project/data/W1.csv", numPixels*hidden, false)
	b1 := loadFlatCSV("mnist_project/data/b1.csv", hidden, false)
	W2 := loadFlatCSV("mnist_project/data/W2.csv", hidden, false)
	b2 := loadFlatCSV("mnist_project/data/b2.csv", 1, false)

	loadTime := time.Since(loadStart)
	fmt.Printf("Loaded %d samples.\n", numSamples)

	// Scratch buffers
	z1Cache := make([]float64, hidden)
	a1Cache := make([]float64, hidden)

	// ─────────────────────────────────────
	//  TRAINING
	// ─────────────────────────────────────
	fmt.Printf("\nTraining neural network: %d -> %d -> 1\n", numPixels, hidden)
	fmt.Printf("Epochs: %d  |  LR: %.3f\n\n", epochs, lr)

	trainStart := time.Now()

	for epoch := 0; epoch < epochs; epoch++ {
		for i := 0; i < numSamples; i++ {
			base := i * (numPixels + 1)

			for h := 0; h < hidden; h++ {
				z := b1[h]
				for p := 0; p < numPixels; p++ {
					z += data[base+p] * W1[p*hidden+h]
				}
				z1Cache[h] = z
				a1Cache[h] = relu(z)
			}

			z2    := b2[0]
			for h := 0; h < hidden; h++ {
				z2 += a1Cache[h] * W2[h]
			}
			a2    := sigmoid(z2)
			label := data[base+numPixels]
			dA2   := a2 - label

			
			for h := 0; h < hidden; h++ {
				W2[h] -= lr * dA2 * a1Cache[h]
			}
			b2[0] -= lr * dA2

			for h := 0; h < hidden; h++ {
				dA1h := dA2 * W2[h] * reluD(z1Cache[h])
				for p := 0; p < numPixels; p++ {
					W1[p*hidden+h] -= lr * dA1h * data[base+p]
				}
				b1[h] -= lr * dA1h
			}
		}
		fmt.Printf("  Epoch %d/%d\n", epoch+1, epochs)
	}

	trainTime := time.Since(trainStart)

	// ─────────────────────────────────────
	//  TESTING
	// ─────────────────────────────────────
	fmt.Printf("\nTesting on all %d samples...\n", numSamples)
	testStart := time.Now()

	correct := 0
	for i := 0; i < numSamples; i++ {
		base := i * (numPixels + 1)

		for h := 0; h < hidden; h++ {
			z := b1[h]
			for p := 0; p < numPixels; p++ {
				z += data[base+p] * W1[p*hidden+h]
			}
			a1Cache[h] = relu(z)
		}

		z2 := b2[0]
		for h := 0; h < hidden; h++ {
			z2 += a1Cache[h] * W2[h]
		}

		pred  := 0.0
		if z2 > 0.0 { pred = 1.0 }
		label := data[base+numPixels]
		if pred == label {
			correct++
		}
	}

	testTime  := time.Since(testStart)
	totalTime := time.Since(totalStart)

	// ─────────────────────────────────────
	//  RESULTS
	// ─────────────────────────────────────
	fmt.Println("\n========================================")
	fmt.Printf("  FINAL ACCURACY: %d / %d\n", correct, numSamples)
	fmt.Printf("  (%.2f%%)\n", 100.0*float64(correct)/float64(numSamples))
	fmt.Println("========================================")
	fmt.Println("\n--- TIMING BREAKDOWN ---")
	fmt.Printf("  Load time:     %f seconds\n", loadTime.Seconds())
	fmt.Printf("  Training time: %f seconds\n", trainTime.Seconds())
	fmt.Printf("  Test time:     %f seconds\n", testTime.Seconds())
	fmt.Printf("  TOTAL time:    %f seconds\n", totalTime.Seconds())
	fmt.Println("------------------------")
}

File: mnist_project/nn_mnist.py
import csv
import math
import time

# ─────────────────────────────────────────
#  CONFIG
# ─────────────────────────────────────────
DATA_PATH = "mnist_project/data/Mnist_Binary_0_vs_1_1000_20x20.csv"
W1_PATH   = "mnist_project/data/W1.csv"
B1_PATH   = "mnist_project/data/b1.csv"
W2_PATH   = "mnist_project/data/W2.csv"
B2_PATH   = "mnist_project/data/b2.csv"

NUM_SAMPLES = 1000
NUM_PIXELS  = 400
HIDDEN      = 64
EPOCHS      = 10
LR          = 0.01

total_start = time.time()

# ─────────────────────────────────────────
#  LOAD FLAT CSV
#  Same strategy as all other languages:
#  stream every value, skip header if present
# ─────────────────────────────────────────
def load_flat_csv(path, count, has_header=False):
    values = []
    with open(path, newline="") as f:
        reader = csv.reader(f)
        if has_header:
            next(reader)
        for row in reader:
            for val in row:
                val = val.strip()
                if val:
                    values.append(float(val))
                if len(values) >= count:
                    return values
    return values

print("Loading dataset and weights...")
load_start = time.time()

# flat: data[i*401 + p] = pixel p,  data[i*401 + 400] = label
data = load_flat_csv(DATA_PATH, NUM_SAMPLES * (NUM_PIXELS + 1), has_header=True)
W1   = load_flat_csv(W1_PATH, NUM_PIXELS * HIDDEN)
b1   = load_flat_csv(B1_PATH, HIDDEN)
W2   = load_flat_csv(W2_PATH, HIDDEN)
b2   = load_flat_csv(B2_PATH, 1)

load_time = time.time() - load_start
print(f"Loaded {NUM_SAMPLES} samples.")

# ─────────────────────────────────────────
#  ACTIVATIONS
# ─────────────────────────────────────────
def relu(x):       return x if x > 0.0 else 0.0
def relu_d(x):     return 1.0 if x > 0.0 else 0.0
def sigmoid(x):
    if x < -500:   return 0.0
    if x >  500:   return 1.0
    return 1.0 / (1.0 + math.exp(-x))

# Scratch buffers — same as C* z1_cache / a1_cache
z1_cache = [0.0] * HIDDEN
a1_cache = [0.0] * HIDDEN

# ─────────────────────────────────────────
#  TRAINING
# ─────────────────────────────────────────
print(f"\nTraining neural network: {NUM_PIXELS} -> {HIDDEN} -> 1")
print(f"Epochs: {EPOCHS}  |  LR: {LR}\n")
train_start = time.time()

for epoch in range(EPOCHS):
    for i in range(NUM_SAMPLES):
        base = i * (NUM_PIXELS + 1)

        # ── FORWARD PASS ──────────────────
        for h in range(HIDDEN):
            z = b1[h]
            for p in range(NUM_PIXELS):
                z += data[base + p] * W1[p * HIDDEN + h]
            z1_cache[h] = z
            a1_cache[h] = relu(z)

        z2    = b2[0]
        for h in range(HIDDEN):
            z2 += a1_cache[h] * W2[h]
        a2    = sigmoid(z2)
        label = data[base + NUM_PIXELS]
        d_a2  = a2 - label

        # ── BACKWARD PASS ─────────────────
        for h in range(HIDDEN):
            W2[h] -= LR * d_a2 * a1_cache[h]
        b2[0] -= LR * d_a2

        for h in range(HIDDEN):
            d_a1_h = d_a2 * W2[h] * relu_d(z1_cache[h])
            for p in range(NUM_PIXELS):
                W1[p * HIDDEN + h] -= LR * d_a1_h * data[base + p]
            b1[h] -= LR * d_a1_h

    print(f"  Epoch {epoch + 1}/{EPOCHS}")

train_time = time.time() - train_start

# ─────────────────────────────────────────
#  TESTING
# ─────────────────────────────────────────
print("\nTesting on all 1000 samples...")
test_start = time.time()

correct = 0
for i in range(NUM_SAMPLES):
    base = i * (NUM_PIXELS + 1)

    for h in range(HIDDEN):
        z = b1[h]
        for p in range(NUM_PIXELS):
            z += data[base + p] * W1[p * HIDDEN + h]
        a1_cache[h] = relu(z)

    z2 = b2[0]
    for h in range(HIDDEN):
        z2 += a1_cache[h] * W2[h]

    pred  = 1.0 if z2 > 0.0 else 0.0
    label = data[base + NUM_PIXELS]
    if pred == label:
        correct += 1

test_time  = time.time() - test_start
total_time = time.time() - total_start

# ─────────────────────────────────────────
#  RESULTS
# ─────────────────────────────────────────
print(f"\n{'='*40}")
print(f"  FINAL ACCURACY: {correct} / {NUM_SAMPLES}")
print(f"  ({100.0 * correct / NUM_SAMPLES:.2f}%)")
print(f"{'='*40}")
print(f"\n--- TIMING BREAKDOWN ---")
print(f"  Load time:     {load_time:.6f} seconds")
print(f"  Training time: {train_time:.6f} seconds")
print(f"  Test time:     {test_time:.6f} seconds")
print(f"  TOTAL time:    {total_time:.6f} seconds")
print(f"------------------------")